desktop/core/src/desktop/js/catalog/optimizer/LocalStrategy.ts

```diff
@@ -0,0 +1,158 @@
+// Licensed to Cloudera, Inc. under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  Cloudera, Inc. licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { CancellablePromise } from 'api/cancellablePromise';
+import contextCatalog from 'catalog/contextCatalog';
+import { OptimizerMeta, TableSourceMeta } from 'catalog/DataCatalogEntry';
+import { TopAggs, TopColumns, TopFilters, TopJoins } from 'catalog/MultiTableEntry';
+import {
+  CompatibilityOptions,
+  MetaOptions,
+  Optimizer,
+  OptimizerRisk,
+  PopularityOptions,
+  RiskOptions,
+  SimilarityOptions
+} from 'catalog/optimizer/optimizer';
+
+import dataCatalog, { OptimizerResponse } from 'catalog/dataCatalog';
+import sqlParserRepository from 'parse/sql/sqlParserRepository';
+import { Connector, Namespace } from 'types/config';
+import I18n from 'utils/i18n';
+
+export default class LocalStrategy implements Optimizer {
```

Comment:
Note for later: we should be able to use both js + api strategy at the same time actually (e.g. help on FKs, top columns with ids ... can always be there via the metadata0

Reply:
It can always be done by adding a new strategy that can proxy certain functions to the ApiStrategy. Alternatively, specify on the connector which function should be generic API.